File: AM_Program/AM_Shared/clsDBStatusLogger.vb
'*********************************************************************************************************
' Written by Matthew Monroe for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
' Copyright 2009, Battelle Memorial Institute
' Created 02/09/2009
'
'*********************************************************************************************************

Option Strict On

Public Class clsDBStatusLogger

#Region "Structures"

	Public Structure udtStatusInfoType
		Public MgrName As String
		Public MgrStatus As IStatusFile.EnumMgrStatus
		Public LastUpdate As DateTime
        Public LastStartTime As DateTime

        ''' <summary>
        ''' Overall CPU utilization of all threads
        ''' </summary>
        ''' <remarks></remarks>
        Public CPUUtilization As Single

        ''' <summary>
        ''' System-wide free memory
        ''' </summary>
        ''' <remarks></remarks>
        Public FreeMemoryMB As Single

        ''' <summary>
        ''' Return the ProcessID of the Analysis manager
        ''' </summary>
        ''' <remarks></remarks>
        Public ProcessID As Integer

        ''' <summary>
        ''' ProcessID of an externally spawned process
        ''' </summary>
        ''' <remarks>0 if no external process running</remarks>
        Public ProgRunnerProcessID As Integer

        ''' <summary>
        ''' Number of cores in use by an externally spawned process
        ''' </summary>
        ''' <remarks></remarks>
        Public ProgRunnerCoreUsage As Single

		Public MostRecentErrorMessage As String
		Public Task As udtTaskInfoType
	End Structure

	Public Structure udtTaskInfoType
		Public Tool As String
		Public Status As IStatusFile.EnumTaskStatus
		Public DurationHours As Single		 ' Task duration, in hours
		Public Progress As Single		' Percent complete, value between 0 and 100
		Public CurrentOperation As String
		Public TaskDetails As udtTaskDetailsType
	End Structure

	Public Structure udtTaskDetailsType
		Public Status As IStatusFile.EnumTaskStatusDetail
		Public Job As Integer
		Public JobStep As Integer
		Public Dataset As String
		Public MostRecentLogMessage As String
		Public MostRecentJobInfo As String
		Public SpectrumCount As Integer
	End Structure

#End Region

#Region "Module variables"

    ''' <summary>
    ''' Stored procedure that could be used to report manager status; typically not used
    ''' </summary>
    ''' <remarks>This stored procedure is valid, but the primary way that we track status is when WriteStatusFile calls LogStatusToMessageQueue</remarks>
	Private Const SP_NAME_UPDATE_MANAGER_STATUS As String = "UpdateManagerAndTaskStatus"

	'Status file name and location
	Private ReadOnly m_DBConnectionString As String

	' The minimum interval between updating the manager status in the database
	Private m_DBStatusUpdateIntervalMinutes As Single = 1

#End Region

#Region "Properties"

	Public ReadOnly Property DBConnectionString() As String
		Get
			Return m_DBConnectionString
		End Get
	End Property

	Public Property DBStatusUpdateIntervalMinutes() As Single
		Get
			Return m_DBStatusUpdateIntervalMinutes
		End Get
        Set(value As Single)
            If value < 0 Then value = 0
            m_DBStatusUpdateIntervalMinutes = value
        End Set
    End Property
#End Region

#Region "Methods"

    ''' <summary>
    ''' Constructor
    ''' </summary>
    ''' <param name="strDBConnectionString">Database connection string</param>
    ''' <param name="sngDBStatusUpdateIntervalMinutes">Minimum interval between updating the manager status in the database</param>
    ''' <remarks></remarks>
    Public Sub New(strDBConnectionString As String, sngDBStatusUpdateIntervalMinutes As Single)
        If strDBConnectionString Is Nothing Then strDBConnectionString = String.Empty
        m_DBConnectionString = strDBConnectionString
        m_DBStatusUpdateIntervalMinutes = sngDBStatusUpdateIntervalMinutes
    End Sub

    ''' <summary>
    ''' Send status information to the database
    ''' </summary>
    ''' <param name="udtStatusInfo"></param>
    ''' <param name="blnForceLogToDB"></param>
    ''' <remarks>This function is valid, but the primary way that we track status is when WriteStatusFile calls LogStatusToMessageQueue</remarks>
    Public Sub LogStatus(udtStatusInfo As udtStatusInfoType, blnForceLogToDB As Boolean)

        Static dtLastWriteTime As DateTime = Date.UtcNow.Subtract(New TimeSpan(1, 0, 0))

        Dim MyConnection As SqlClient.SqlConnection
        Dim myCmd As New SqlClient.SqlCommand

        Try
            If String.IsNullOrEmpty(m_DBConnectionString) Then
                ' Connection string not defined; unable to continue
                Exit Sub
            End If

            If Not blnForceLogToDB AndAlso Date.UtcNow.Subtract(dtLastWriteTime).TotalMinutes < m_DBStatusUpdateIntervalMinutes Then
                ' Not enough time has elapsed since the last write; exit sub
                Exit Sub
            End If
            dtLastWriteTime = Date.UtcNow


            MyConnection = New SqlClient.SqlConnection(m_DBConnectionString)
            MyConnection.Open()

            'Set up the command object prior to SP execution
            With myCmd

                .CommandType = CommandType.StoredProcedure
                .CommandText = SP_NAME_UPDATE_MANAGER_STATUS
                .Connection = MyConnection

                .Parameters.Add(New SqlClient.SqlParameter("@Return", SqlDbType.Int))
                .Parameters.Item("@Return").Direction = ParameterDirection.ReturnValue


                ' Manager items
                AddSPParameter(.Parameters, "@MgrName", udtStatusInfo.MgrName, 128)
                AddSPParameter(.Parameters, "@MgrStatusCode", udtStatusInfo.MgrStatus)

                AddSPParameter(.Parameters, "@LastUpdate", udtStatusInfo.LastUpdate.ToLocalTime())
                AddSPParameter(.Parameters, "@LastStartTime", udtStatusInfo.LastStartTime.ToLocalTime())
                AddSPParameter(.Parameters, "@CPUUtilization", udtStatusInfo.CPUUtilization)
                AddSPParameter(.Parameters, "@FreeMemoryMB", udtStatusInfo.FreeMemoryMB)
                AddSPParameter(.Parameters, "@ProcessID", udtStatusInfo.ProcessID)
                AddSPParameter(.Parameters, "@ProgRunnerProcessID", udtStatusInfo.ProgRunnerProcessID)
                AddSPParameter(.Parameters, "@ProgRunnerCoreUsage", udtStatusInfo.ProgRunnerCoreUsage)

                AddSPParameter(.Parameters, "@MostRecentErrorMessage", udtStatusInfo.MostRecentErrorMessage, 1024)

                ' Task items
                AddSPParameter(.Parameters, "@StepTool", udtStatusInfo.Task.Tool, 128)
                AddSPParameter(.Parameters, "@TaskStatusCode", udtStatusInfo.Task.Status)
                AddSPParameter(.Parameters, "@DurationHours", udtStatusInfo.Task.DurationHours)
                AddSPParameter(.Parameters, "@Progress", udtStatusInfo.Task.Progress)
                AddSPParameter(.Parameters, "@CurrentOperation", udtStatusInfo.Task.CurrentOperation, 256)

                ' Task detail items
                AddSPParameter(.Parameters, "@TaskDetailStatusCode", udtStatusInfo.Task.TaskDetails.Status)
                AddSPParameter(.Parameters, "@Job", udtStatusInfo.Task.TaskDetails.Job)
                AddSPParameter(.Parameters, "@JobStep", udtStatusInfo.Task.TaskDetails.JobStep)
                AddSPParameter(.Parameters, "@Dataset", udtStatusInfo.Task.TaskDetails.Dataset, 256)
                AddSPParameter(.Parameters, "@MostRecentLogMessage", udtStatusInfo.Task.TaskDetails.MostRecentLogMessage, 1024)
                AddSPParameter(.Parameters, "@MostRecentJobInfo", udtStatusInfo.Task.TaskDetails.MostRecentJobInfo, 256)
                AddSPParameter(.Parameters, "@SpectrumCount", udtStatusInfo.Task.TaskDetails.SpectrumCount)

                AddSPParameterOutput(.Parameters, "@message", String.Empty, 512)

            End With

            'Execute the SP
            myCmd.ExecuteNonQuery()

        Catch ex As Exception
            ' Ignore errors here
            Console.WriteLine("Error in clsDBStatusLogger.LogStatus: " & ex.Message)
        End Try

    End Sub

    Private Sub AddSPParameter(ByRef objParameters As SqlClient.SqlParameterCollection, strParamName As String, strValue As String, intVarCharLength As Integer)
        ' Make sure the parameter starts with an @ sign
        If Not strParamName.StartsWith("@") Then
            strParamName = "@" & strParamName
        End If

        objParameters.Add(New SqlClient.SqlParameter(strParamName, SqlDbType.VarChar, intVarCharLength))
        objParameters.Item(strParamName).Direction = ParameterDirection.Input
        objParameters.Item(strParamName).Value = strValue
    End Sub

    Private Sub AddSPParameter(ByRef objParameters As SqlClient.SqlParameterCollection, strParamName As String, intValue As Integer)
        ' Make sure the parameter starts with an @ sign
        If Not strParamName.StartsWith("@") Then
            strParamName = "@" & strParamName
        End If

        objParameters.Add(New SqlClient.SqlParameter(strParamName, SqlDbType.Int))
        objParameters.Item(strParamName).Direction = ParameterDirection.Input
        objParameters.Item(strParamName).Value = intValue
    End Sub

    Private Sub AddSPParameter(ByRef objParameters As SqlClient.SqlParameterCollection, strParamName As String, dtValue As DateTime)
        ' Make sure the parameter starts with an @ sign
        If Not strParamName.StartsWith("@") Then
            strParamName = "@" & strParamName
        End If

        objParameters.Add(New SqlClient.SqlParameter(strParamName, SqlDbType.DateTime))
        objParameters.Item(strParamName).Direction = ParameterDirection.Input
        objParameters.Item(strParamName).Value = dtValue
    End Sub

    Private Sub AddSPParameter(ByRef objParameters As SqlClient.SqlParameterCollection, strParamName As String, sngValue As Single)
        ' Make sure the parameter starts with an @ sign
        If Not strParamName.StartsWith("@") Then
            strParamName = "@" & strParamName
        End If

        objParameters.Add(New SqlClient.SqlParameter(strParamName, SqlDbType.Real))
        objParameters.Item(strParamName).Direction = ParameterDirection.Input
        objParameters.Item(strParamName).Value = sngValue
    End Sub

    Private Sub AddSPParameterOutput(ByRef objParameters As SqlClient.SqlParameterCollection, strParamName As String, strValue As String, intVarCharLength As Integer)
        ' Make sure the parameter starts with an @ sign
        If Not strParamName.StartsWith("@") Then
            strParamName = "@" & strParamName
        End If

        objParameters.Add(New SqlClient.SqlParameter(strParamName, SqlDbType.VarChar, intVarCharLength))
        objParameters.Item(strParamName).Direction = ParameterDirection.Output
        objParameters.Item(strParamName).Value = strValue
    End Sub

#End Region

End Class

File: AM_Program/AM_Shared/clsDBStatusLogger.cs
using System;
using System.Data;
using PRISM;
using PRISMDatabaseUtils;

//*********************************************************************************************************
// Written by Matthew Monroe for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
// Copyright 2009, Battelle Memorial Institute
// Created 02/09/2009
//
//*********************************************************************************************************


namespace AnalysisManagerBase
{

    /// <summary>
    /// Status logger
    /// </summary>
    public class clsDBStatusLogger : EventNotifier
    {

        #region "Structures"

        /// <summary>
        /// Status info
        /// </summary>
        public struct udtStatusInfoType
        {
            /// <summary>
            /// Manager name
            /// </summary>
            public string MgrName;

            /// <summary>
            /// Manager status
            /// </summary>
            public EnumMgrStatus MgrStatus;

            /// <summary>
            /// Last update time (UTC-based)
            /// </summary>
            public DateTime LastUpdate;

            /// <summary>
            /// Last start time (UTC-based)
            /// </summary>
            public DateTime LastStartTime;

            /// <summary>
            /// Overall CPU utilization of all threads
            /// </summary>
            /// <remarks></remarks>
            public float CPUUtilization;

            /// <summary>
            /// System-wide free memory
            /// </summary>
            /// <remarks></remarks>
            public float FreeMemoryMB;

            /// <summary>
            /// Return the ProcessID of the Analysis manager
            /// </summary>
            /// <remarks></remarks>
            public int ProcessID;

            /// <summary>
            /// ProcessID of an externally spawned process
            /// </summary>
            /// <remarks>0 if no external process running</remarks>
            public int ProgRunnerProcessID;

            /// <summary>
            /// Number of cores in use by an externally spawned process
            /// </summary>
            /// <remarks></remarks>
            public float ProgRunnerCoreUsage;

            /// <summary>
            /// Most recent error message
            /// </summary>
            public string MostRecentErrorMessage;

            /// <summary>
            /// Current task
            /// </summary>
            public udtTaskInfoType Task;
        }

        /// <summary>
        /// Task info
        /// </summary>
        public struct udtTaskInfoType
        {
            /// <summary>
            /// Analysis tool name
            /// </summary>
            public string Tool;

            /// <summary>
            /// Task status
            /// </summary>
            public EnumTaskStatus Status;

            /// <summary>
            /// Task duration, in hours
            /// </summary>
            public float DurationHours;

            /// <summary>
            /// Percent complete, value between 0 and 100
            /// </summary>
            public float Progress;

            /// <summary>
            /// Current operation
            /// </summary>
            public string CurrentOperation;

            /// <summary>
            /// Task details
            /// </summary>
            public udtTaskDetailsType TaskDetails;
        }

        /// <summary>
        /// Task details
        /// </summary>
        public struct udtTaskDetailsType
        {
            /// <summary>
            /// Task status detail
            /// </summary>
            public EnumTaskStatusDetail Status;

            /// <summary>
            /// Job number
            /// </summary>
            public int Job;

            /// <summary>
            /// Job step
            /// </summary>
            public int JobStep;

            /// <summary>
            /// Dataset name
            /// </summary>
            public string Dataset;

            /// <summary>
            /// Most recent log message
            /// </summary>
            public string MostRecentLogMessage;

            /// <summary>
            /// Most recent job info
            /// </summary>
            public string MostRecentJobInfo;

            /// <summary>
            /// Number of spectra in the instrument data file
            /// </summary>
            /// <remarks>Only used by certain tools (e.g. SEQUEST, X!Tandem, and Inspect)</remarks>
            public int SpectrumCount;
        }

        #endregion

        #region "Module variables"

        /// <summary>
        /// Stored procedure that could be used to report manager status; typically not used
        /// </summary>
        /// <remarks>This stored procedure is valid, but the primary way that we track status is when WriteStatusFile calls LogStatusToMessageQueue</remarks>
        private const string SP_NAME_UPDATE_MANAGER_STATUS = "UpdateManagerAndTaskStatus";

        /// <summary>
        /// The minimum interval between updating the manager status in the database
        /// </summary>
        private float mDBStatusUpdateIntervalMinutes;

        private DateTime mLastWriteTime;

        #endregion

        #region "Properties"

        /// <summary>
        /// Database connection string
        /// </summary>
        public string DBConnectionString { get; }

        /// <summary>
        /// Status update interval, in minutes
        /// </summary>
        public float DBStatusUpdateIntervalMinutes
        {
            get => mDBStatusUpdateIntervalMinutes;
            set
            {
                if (value < 0)
                    value = 0;
                mDBStatusUpdateIntervalMinutes = value;
            }
        }

        /// <summary>
        /// Pipeline database stored procedure executor
        /// </summary>
        private IDBTools PipelineDBProcedureExecutor { get; }

        #endregion

        #region "Methods"

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="dbConnectionString">Database connection string</param>
        /// <param name="dbStatusUpdateIntervalMinutes">Minimum interval between updating the manager status in the database</param>
        /// <remarks></remarks>
        public clsDBStatusLogger(string dbConnectionString, float dbStatusUpdateIntervalMinutes)
        {
            if (dbConnectionString == null)
                dbConnectionString = string.Empty;

            PipelineDBProcedureExecutor = DbToolsFactory.GetDBTools(dbConnectionString);
            RegisterEvents(PipelineDBProcedureExecutor);

            DBConnectionString = dbConnectionString;
            mDBStatusUpdateIntervalMinutes = dbStatusUpdateIntervalMinutes;
            mLastWriteTime = DateTime.MinValue;
        }

        /// <summary>
        /// Send status information to the database
        /// </summary>
        /// <param name="statusInfo"></param>
        /// <param name="forceLogToDB"></param>
        /// <remarks>This function is valid, but the primary way that we track status is when WriteStatusFile calls LogStatusToMessageQueue</remarks>
        public void LogStatus(udtStatusInfoType statusInfo, bool forceLogToDB)
        {

            try
            {
                if (string.IsNullOrEmpty(DBConnectionString))
                {
                    // Connection string not defined; unable to continue
                    return;
                }

                if (!forceLogToDB && DateTime.UtcNow.Subtract(mLastWriteTime).TotalMinutes < mDBStatusUpdateIntervalMinutes)
                {
                    // Not enough time has elapsed since the last write; exit method
                    return;
                }
                mLastWriteTime = DateTime.UtcNow;

                // Set up the command object prior to SP execution
                var cmd = PipelineDBProcedureExecutor.CreateCommand(SP_NAME_UPDATE_MANAGER_STATUS, CommandType.StoredProcedure);

                // Manager items
                AddSPParameter(cmd, "@MgrName", statusInfo.MgrName, 128);
                AddSPParameter(cmd, "@MgrStatusCode", (int)statusInfo.MgrStatus);

                AddSPParameter(cmd, "@LastUpdate", statusInfo.LastUpdate.ToLocalTime());
                AddSPParameter(cmd, "@LastStartTime", statusInfo.LastStartTime.ToLocalTime());
                AddSPParameter(cmd, "@CPUUtilization", statusInfo.CPUUtilization);
                AddSPParameter(cmd, "@FreeMemoryMB", statusInfo.FreeMemoryMB);
                AddSPParameter(cmd, "@ProcessID", statusInfo.ProcessID);
                AddSPParameter(cmd, "@ProgRunnerProcessID", statusInfo.ProgRunnerProcessID);
                AddSPParameter(cmd, "@ProgRunnerCoreUsage", statusInfo.ProgRunnerCoreUsage);

                AddSPParameter(cmd, "@MostRecentErrorMessage", statusInfo.MostRecentErrorMessage, 1024);

                // Task items
                AddSPParameter(cmd, "@StepTool", statusInfo.Task.Tool, 128);
                AddSPParameter(cmd, "@TaskStatusCode", (int)statusInfo.Task.Status);
                AddSPParameter(cmd, "@DurationHours", statusInfo.Task.DurationHours);
                AddSPParameter(cmd, "@Progress", statusInfo.Task.Progress);
                AddSPParameter(cmd, "@CurrentOperation", statusInfo.Task.CurrentOperation, 256);

                // Task detail items
                AddSPParameter(cmd, "@TaskDetailStatusCode", (int)statusInfo.Task.TaskDetails.Status);
                AddSPParameter(cmd, "@Job", statusInfo.Task.TaskDetails.Job);
                AddSPParameter(cmd, "@JobStep", statusInfo.Task.TaskDetails.JobStep);
                AddSPParameter(cmd, "@Dataset", statusInfo.Task.TaskDetails.Dataset, 256);
                AddSPParameter(cmd, "@MostRecentLogMessage", statusInfo.Task.TaskDetails.MostRecentLogMessage, 1024);
                AddSPParameter(cmd, "@MostRecentJobInfo", statusInfo.Task.TaskDetails.MostRecentJobInfo, 256);
                AddSPParameter(cmd, "@SpectrumCount", statusInfo.Task.TaskDetails.SpectrumCount);

                AddSPParameterOutput(cmd, "@message", string.Empty, 512);

                // Execute the SP
                PipelineDBProcedureExecutor.ExecuteSP(cmd);

            }
            catch (Exception ex)
            {
                // Ignore errors here
                Console.WriteLine("Error in clsDBStatusLogger.LogStatus: " + ex.Message);
            }
        }

        private void AddSPParameter(System.Data.Common.DbCommand cmd, string paramName, string value, int varCharLength)
        {
            // Make sure the parameter starts with an @ sign
            if (!paramName.StartsWith("@"))
            {
                paramName = "@" + paramName;
            }

            PipelineDBProcedureExecutor.AddParameter(cmd, paramName, SqlType.VarChar, varCharLength, value);
        }

        private void AddSPParameter(System.Data.Common.DbCommand cmd, string paramName, int value)
        {
            // Make sure the parameter starts with an @ sign
            if (!paramName.StartsWith("@"))
            {
                paramName = "@" + paramName;
            }

            PipelineDBProcedureExecutor.AddParameter(cmd, paramName, SqlType.Int).Value = value;
        }

        private void AddSPParameter(System.Data.Common.DbCommand cmd, string paramName, DateTime value)
        {
            // Make sure the parameter starts with an @ sign
            if (!paramName.StartsWith("@"))
            {
                paramName = "@" + paramName;
            }

            PipelineDBProcedureExecutor.AddParameter(cmd, paramName, SqlType.DateTime).Value = value;
        }

        private void AddSPParameter(System.Data.Common.DbCommand cmd, string paramName, float value)
        {
            // Make sure the parameter starts with an @ sign
            if (!paramName.StartsWith("@"))
            {
                paramName = "@" + paramName;
            }

            PipelineDBProcedureExecutor.AddParameter(cmd, paramName, SqlType.Real).Value = value;
        }

        private void AddSPParameterOutput(System.Data.Common.DbCommand cmd, string paramName, string value, int varCharLength)
        {
            // Make sure the parameter starts with an @ sign
            if (!paramName.StartsWith("@"))
            {
                paramName = "@" + paramName;
            }

            PipelineDBProcedureExecutor.AddParameter(cmd, paramName, SqlType.VarChar, varCharLength, value, ParameterDirection.Output);
        }

        #endregion

    }
}
